rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,267 @@
+  "agent_update_jitter_seconds": 10,
+}
+```
+Notes:
+- Critical updates are achieved by serving `agent_update_after` with the current time.
+- The Teleport proxy translates upgrade hours (below) into a specific time after which all agents should be upgraded.
+- If an agent misses an upgrade window, it will always update immediately.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: on|off
+  # agent_update_hour sets the hour in UTC at which clients should update their agents.
+  # The value -1 will set the upgrade time to the current time, resulting in immediate upgrades.
+  agent_update_hour: -1-23
+  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
+  # The agent upgrader will pick a random time within this duration in which to upgrade.
+  agent_update_jitter_seconds: 0-MAXINT64
+  
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-auto-update=off
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-hour=3
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-jitter-seconds=600
+Automatic updates configuration has been updated.
+```
+
+```yaml
+kind: autoupdate_version
+spec:
+  # agent_version is the version of the agent the cluster will advertise.
+  # Can be auto (match the version of the proxy) or an exact semver formatted
+  # version.
+  agent_version: auto|X.Y.Z
+
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-version=15.1.1
+Automatic updates configuration has been updated.
+```
+
+Notes:
+- These two resources are separate so that Cloud customers can be restricted from updating `autoupdate_version`, while maintaining control over the rollout.
+
+### Filesystem
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport-updater -> /var/lib/teleport/versions/15.0.0/bin/teleport-updater
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+proxy: mytenant.teleport.sh
+enabled: true
+active_version: 15.1.1
+```
+
+### Runtime
```

Comment:
> We need to make sure the teleport-updater package can support OSS, Enterprise, and Enterprise FIPs installations.

I think we can put this information into the `ping` endpoint. Then you no longer have to worry about which binary you need to install (OSS, Enterprise, FIPS Enterprise), you jus install the `teleport-updater` and it downloads and runs this right version.

Reply:
Something we should consider: the `apt-get install && teleport-updater enable` approach works fine if you running the commands on the host (or running something like Ansible), what would we recommend to someone that builds an AMI using this new install method?

Reply:
I think I've covered everything above:
- Added new "Installers" section with scripts changes and VM/container image instructions
- Added new "Documentation" section with links and notes
- New `server_edition` field is added to `ping`, along with associated changes to "Runtime" section logic

Let me know if this looks good.

Reply:
If we make this new installer the default for everything (which I think is a great approach, we must reduce complexity) I would suggest going further by making the new binary the only recommended way of installing teleport, even without Automatic Updates.

We could name it `teleport-version-manager` or whatever name that doesn't imply it requires AUs and expose the following commands:
- `tvm update` as already described in the RFD
- `tvm follow` to follow updates from a proxy, this is the current RFD `enable`
- `tvm install <X.Y.Z|teleport.example.com>` would be an additional command not enabling AUs, but installing a static version (given, or from a proxy).

I think this would answer
> what would we recommend to someone that builds an AMI using this new install method

We would recommend "install tvm, and run `tvm install teleport.example.com` or `tvm follow teleport.example.com` on startup".

Reply:
I like your design, proposed commands, and shorter name. Two minor concerns:
- `teleport-version-manager` might suggest that the binary manages N number of installed Teleport versions (like rvm, nvm, etc.), but `teleport-updater` only permits two versions to be installed at any given time (active and backup). I'm not sure if users will understand that `tvm install` also uninstalls older versions.
- Renaming the package would add complexity to the rollout. While it's possible to rename packages without user interaction in most distros, I wonder if we could defer the rename until after this is shipped to reduce the number of changes. That said, while we can defer renaming the package, we should get consensus on the binary name before we ship.

I have no strong opinion about these names. If other reviewers like `tvm`/`teleport-version-manager`, happy to change it.

> making the new binary the only recommended way of installing teleport, even without Automatic Updates.

I don't think we should stop publishing fixed-version packages for Teleport, but I definitely agree we should recommend that all agents use this method.

Reply:
> Renaming the package would add complexity to the rollout.

:+1: the existing package should be reused. In the worst-case scenario, we keep the package name and change the binary name (and maybe drop a script named `teleport-updater` that just tells users to call `tvm` instead). This way, existing users would not attempt to use old commands with the new binary.

> teleport-version-manager might suggest that the binary manages N number of installed Teleport versions (like rvm, nvm, etc.), but teleport-updater only permits two versions to be installed at any given time (active and backup). I'm not sure if users will understand that tvm install also uninstalls older version

Would `tvm use-version vX.Y.Z` or `tvm use vX.Y.Z` be better?

Reply:
> Would tvm use-version vX.Y.Z or tvm use vX.Y.Z be better?

`tvm use` sounds better to me, but I don't have a strong opinion.

Happy to rename. I'll wait for a few others to chime in, and change it when we some consensus -- names can be contentious 🙂

Reply:
👍 Works for me. Maybe there are two phases here:
- Ship this RFD, which minimizes user-facing changes
- Re-design UX, simplify scripts, etc.

Reply:
If we postpone the simplification we must absolutely ensure we do it the next quart.

Because the installation methods are already a mess and we're adding new methods and scripts. This is not maintainable and self-hosted customers are already complaining about how hard it is to just install teleport, and this is not an issue we can solve with docs. We've tried for the last year and the installation docs only become worse and more obscure since I joined Teleport.

I'm against adding new ways without removing the old ones but I feel like I will not have the final word here.